# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension or as a self-invoking function.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
def add_two(x):
    return x + 2

print(add_two(4))

6


In [2]:
# Written as a Lambda function
# Self-invoking function - meaning it immediately calls itself
(lambda x: x + 3)(4)

7

#### Saving to a Variable

In [3]:
add_three = lambda x: x + 3

print(add_three)

<function <lambda> at 0x0000022AED562790>


In [4]:
add_three(4)

7

#### Multiple Inputs

In [5]:
# Multiple inputs with no variables
(lambda x,y,z: x * y / z)(6, 4, 2)

12.0

In [6]:
# Multiple inputs with a variable attached
mult_test = lambda x, y, z: x * y / z
mult_test(8, 6, 4)

12.0

#### Passing a Lambda into a Function

In [7]:
def multiply(f, num):
    """
    f expects a function that accepts one argument
    num expects a number
    """
    print(f)
    num -= 50
    return f(num)

multiply(lambda x: (x + 5)/10 , 100)

<function <lambda> at 0x0000022AED595040>


5.5

In [8]:
abs(-100)

100

In [9]:
help(abs)

Help on built-in function abs in module builtins:

abs(x, /)
    Return the absolute value of the argument.



#### Returning a Lambda from a Function

In [10]:
# Regularly defined Function
def multiply_test(num):
    return num * 4

my_return = multiply_test(5)
print(type(my_return))

<class 'int'>


In [11]:
# Function within a function
def parent_func(parent_name):
    def first_child(child_name):
        return f'I am {child_name}, a child of {parent_name}'
    return first_child

a_child = parent_func('Mike')

print(a_child)
a_child('Brian')

<function parent_func.<locals>.first_child at 0x0000022AED5963A0>


'I am Brian, a child of Mike'

In [12]:
a_child('Julie')

'I am Julie, a child of Mike'

In [13]:
def return_lambda(b, c):
    return lambda a, d: a * b + c * d

my_func = return_lambda(5, 6)
my_func(2, 3)

28

#### If Statements within Lambdas

In [14]:
def some_condition(num):
    if num > 10:
        return num * 2
    else:
        return num + 2

# lambda *input*: *return when True* if *condition* else *return when false*
lambda_condition = lambda num: num * 2 if num > 10 else num + 2

lambda_condition(18)

36

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the argument passed in and assign the anonymous function to a variable 'f'.</p>

In [15]:
f = lambda num: num**3

f(10)

1000

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [16]:
help(map)

Help on class map in module builtins:

class map(object)
 |  map(func, *iterables) --> map object
 |  
 |  Make an iterator that computes the function using arguments from
 |  each of the iterables.  Stops when the shortest iterable is exhausted.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



In [17]:
def returns_double(num):
    return num * 2

my_list = [4, 7, 2, 5, 8, 3, 6, 2, 6, 2, 6, 5]
print(len(my_list))

list(map(returns_double, my_list))

12


[8, 14, 4, 10, 16, 6, 12, 4, 12, 4, 12, 10]

In [18]:
my_dict = {
    'a': 10,
    'b': 'Hello',
    'c': 20
}

def something(key):
    if isinstance(my_dict[key], int):
        return my_dict[key] ** 2
    else:
        return key

list(map(something, my_dict))

[100, 'b', 400]

In [19]:
my_list_iter = my_list.__iter__()

In [20]:
my_list_iter.__next__()

4

In [21]:
list1 = [1, 2, 3, 4, 5]
list2 = [10,20,30]

def my_func(num1, num2):
    if num1 * num2 > 15:
        return 'Hello'
    else:
        return 'Goodbye'
    
    
list(map(my_func, list1, list2))

['Goodbye', 'Hello', 'Hello']

#### Using Lambda's with Map

In [22]:
my_list = [4, 7, 2, 5, 8, 3, 6, 2, 6, 2, 6, 5]
my_list

[4, 7, 2, 5, 8, 3, 6, 2, 6, 2, 6, 5]

In [23]:
list(map(lambda x: x**2/2, my_list))

[8.0, 24.5, 2.0, 12.5, 32.0, 4.5, 18.0, 2.0, 18.0, 2.0, 18.0, 12.5]

In [24]:
new_map = map(lambda x: x**2/2, my_list)

In [25]:
for i in new_map:
    print(i)

8.0
24.5
2.0
12.5
32.0
4.5
18.0
2.0
18.0
2.0
18.0
12.5


In [26]:
def do_something(x):
    return x**2/2
[(lambda a: a**2/2)(x) for x in my_list]

[8.0, 24.5, 2.0, 12.5, 32.0, 4.5, 18.0, 2.0, 18.0, 2.0, 18.0, 12.5]

In [27]:
list(map(do_something, my_list))

[8.0, 24.5, 2.0, 12.5, 32.0, 4.5, 18.0, 2.0, 18.0, 2.0, 18.0, 12.5]

In [28]:
list(map(lambda num1, num2: 'Hello' if num1 * num2 > 15 else 'Goodbye', list1, list2))

['Goodbye', 'Hello', 'Hello']

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [29]:
a_list = [10, 35, 66, 3, 6, 34, 22, 13, 87]

# [19, 69, 131, 5, 11, 67, 43, 25, 173]
list(map(lambda a: a*2-1, a_list))

[19, 69, 131, 5, 11, 67, 43, 25, 173]

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [30]:
help(filter)

Help on class filter in module builtins:

class filter(object)
 |  filter(function or None, iterable) --> filter object
 |  
 |  Return an iterator yielding those items of iterable for which function(item)
 |  is true. If function is None, return the items that are true.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



In [31]:
def is_even(num):
    if num % 2 == 0:
        return True
    else:
        return False
    
    
list_b = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

list(filter(is_even, list_b))

[2, 4, 6, 8, 10]

#### Using Lambda's with Filter()

In [32]:
def a_names(name):
    return name[0] == 'A'

a_names('Andrew')

True

In [33]:
name = 'Adam'

(name[0] == 'A') == True

True

In [34]:
names = ['Bob', 'Andrew', 'Annie', 'Max', 'Evan', 'Alex']

a_names = list(filter(lambda name: name[0] == 'A', names))
print(a_names)

['Andrew', 'Annie', 'Alex']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [35]:
a_list = [41, 48, 92, 77, 5, 81, 82, 72, 68, 57, 70, 84, 56, 91, 38, 5, 29, 24, 83, 94]
print(a_list)

[41, 48, 92, 77, 5, 81, 82, 72, 68, 57, 70, 84, 56, 91, 38, 5, 29, 24, 83, 94]


In [36]:
import statistics

mean_of_list = statistics.mean(a_list)

def is_greater_than_mean(num):
    return num > mean_of_list


is_greater_than_mean(92)

greater_than = []
for number in a_list:
    if is_greater_than_mean(number):
        greater_than.append(number)

print(greater_than)

[92, 77, 81, 82, 72, 68, 70, 84, 91, 83, 94]


In [41]:
list(filter(lambda num: num > sum(a_list)/len(a_list), a_list))

[92, 77, 81, 82, 72, 68, 70, 84, 91, 83, 94]

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [43]:
# reduce(function, iterable)
from functools import reduce

help(reduce)

Help on built-in function reduce in module _functools:

reduce(...)
    reduce(function, sequence[, initial]) -> value
    
    Apply a function of two arguments cumulatively to the items of a sequence,
    from left to right, so as to reduce the sequence to a single value.
    For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) calculates
    ((((1+2)+3)+4)+5).  If initial is present, it is placed before the items
    of the sequence in the calculation, and serves as a default when the
    sequence is empty.



In [51]:
list1 = [1, 2, 3, 4, 5]

def mult_nums(num1, num2):
    return num1 * num2

reduce(mult_nums, list1)

120

In [48]:
total = 1
for num in list1:
    total *= num
print(total)

120


#### Using Lambda's with Reduce()

In [ ]:
list1 = [1, 2, 3, 4, 5]

reduce(lambda x, y: x*y, list1)

#### In-Class Example #4 <br>
<p>Using reduce to find the max of a list.</p>

In [53]:
import random
random_list = [random.randint(1,100) for i in range(20)]
print(random_list)

[85, 93, 70, 88, 10, 75, 92, 97, 36, 77, 8, 91, 22, 96, 85, 14, 82, 14, 13, 46]


In [55]:
def my_max_func(current_max, potential_new_max):
    if current_max > potential_new_max:
        new_max = current_max
    else:
        new_max = potential_new_max
    return new_max
    
    
reduce(my_max_func, random_list)

97

In [57]:
# Use reduce to find the min using lambda
reduce(lambda a,b: a if a<b else b, random_list)

8

In [58]:
current_max = random_list[0]
for num in random_list:
    if num > current_max:
        current_max = num
print(current_max)

97


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [64]:
def another_func(a_num):
    print('Another Func is running')
    return f'The number is {a_num}'

def my_func(num):
    if num > 15:
        return num
    print(f'My number is {num}')
    num += 1
    return my_func(num)
  
print(f'The return value of my_func when I pass in 10 is {my_func(10)}')

My number is 10
My number is 11
My number is 12
My number is 13
My number is 14
My number is 15
The return value of my_func when I pass in 10 is 16


In [71]:
def add_nums(num):
    # Set base case for recursive function
    if num <= 1:
        print("add_nums(1) = 1")
        return num
    else:
        print(f'add_nums({num}) = {num} + add_nums({num - 1})')
        return num + add_nums(num - 1)
    
add_nums(5)

add_nums(5) = 5 + add_nums(4)
add_nums(4) = 4 + add_nums(3)
add_nums(3) = 3 + add_nums(2)
add_nums(2) = 2 + add_nums(1)
add_nums(1) = 1


15

#### Writing a Factorial Function

In [72]:
# 5! = 5 * 4!
# 4! = 4 * 3!
# 3! = 3 * 2!
# 2! = 2 * 1!
# 1! = 1

def factorial(num):
    if num <= 1:
        return num
    return num * factorial(num-1)

factorial(5)

120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [89]:
def subtract_nums(num):
    # Set base case for recursive function
    if num <= 1:
        print("subtract_nums(1) = 1")
        return num
    else:
        print(f'subtract_nums({num}) = {num} - subtract_nums({num - 1})')
        return num - subtract_nums(num - 1)
    
subtract_nums(10)

subtract_nums(10) = 10 - subtract_nums(9)
subtract_nums(9) = 9 - subtract_nums(8)
subtract_nums(8) = 8 - subtract_nums(7)
subtract_nums(7) = 7 - subtract_nums(6)
subtract_nums(6) = 6 - subtract_nums(5)
subtract_nums(5) = 5 - subtract_nums(4)
subtract_nums(4) = 4 - subtract_nums(3)
subtract_nums(3) = 3 - subtract_nums(2)
subtract_nums(2) = 2 - subtract_nums(1)
subtract_nums(1) = 1


5

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [95]:
def my_range(stop, start, step = 2):
    while start < stop:
        yield start
        start += step

for i in my_range(50, 3):
    print(i)

3
5
7
9
11
13
15
17
19
21
23
25
27
29
31
33
35
37
39
41
43
45
47
49


#### Infinite Generator

In [107]:
# bad, never create infinite loops
def infinite_generator():
    num = 0
    while True:
        yield num
        num += 1
    

In [109]:
forever_gen = infinite_generator()

In [130]:
next(forever_gen)

20

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [131]:
def your_iterator(num):
    while num > 0:
        yield num ** 2
        num -= 1
        

for i in your_iterator(5):
    print(i)
       


25
16
9
4
1


In [147]:
my_iter = your_iterator(10)

In [150]:
next(my_iter)

64

In [151]:
for x in my_iter:
    print(x)

49
36
25
16
9
4
1


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [ ]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]



### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [152]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]



Help on method_descriptor:

sort(self, /, *, key=None, reverse=False)
    Sort the list in ascending order and return None.
    
    The sort is in-place (i.e. the list itself is modified) and stable (i.e. the
    order of two equal elements is maintained).
    
    If a key function is given, apply it once to each list item and sort them,
    ascending or descending, according to their function values.
    
    The reverse flag can be set to sort in descending order.



### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angeles', 111.2), ('Miami', 84.2)]
`

In [153]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angeles",44),("Miami",29)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`